app/controllers/api/account.php

```diff
@@ -3947,21 +3968,24 @@
             throw new Exception(Exception::USER_INVALID_TOKEN);
         }
 
-        $provider = $challenge->getAttribute('provider');
-        $success = match ($provider) {
-            'totp' => Challenge\TOTP::challenge($challenge, $user, $otp),
-            'phone' => Challenge\Phone::challenge($challenge, $user, $otp),
-            'email' => Challenge\Email::challenge($challenge, $user, $otp),
+        $type = $challenge->getAttribute('type');
+
+        $success = match ($type) {
+            Type::TOTP => Challenge\TOTP::challenge($challenge, $user, $otp),
+            Type::PHONE => Challenge\Phone::challenge($challenge, $user, $otp),
+            Type::EMAIL => Challenge\Email::challenge($challenge, $user, $otp),
             default => false
         };
 
-    if (!$success && $provider === 'totp') {
-        $backups = $user->getAttribute('totpBackup', []);
-        if (in_array($otp, $backups)) {
+    if (!$success && $type === Type::TOTP) {
+        $authenticator = TOTP::getAuthenticatorFromUser($user);
+        $data = $authenticator->getAttribute('data', []);
+        if (in_array($otp, $data['backups'])) {
```

Comment:
I know it's unlikely to happen, but we should probably do some sort of null check here just in case. When static analyzer is enabled, it'll complain if we don't.

```suggestion
        $data = $authenticator?->getAttribute('data', []);
        if (in_array($otp, $data['backups'] ?? [])) {
```